private/pkg/dag/dag.go

```diff
@@ -93,11 +93,57 @@ func (g *Graph[Key]) ContainsNode(key Key) bool {
 	return ok
 }
 
-// Walk visits each edge in the Graph starting at the source keys.
+// NumNodes returns the number of nodes in the graph.
+func (g *Graph[Key]) NumNodes() int {
+	g.init()
+	return len(g.keys)
+}
+
+// NumNodes returns the number of edges in the graph.
+func (g *Graph[Key]) NumEdges() int {
+	g.init()
+	var numEdges int
+	for _, node := range g.keyToNode {
+		numEdges += len(node.outboundEdges)
+	}
+	return numEdges
+}
+
+// WalkNodes visited each node in the Graph based on insertion order.
+//
+// f is called for each node. The first argument is the key for the node,
+// the second argument is all inbound edges, the third argument
+// is all outbound edges.
+func (g *Graph[Key]) WalkNodes(f func(Key, []Key, []Key) error) error {
+	g.init()
+	for _, key := range g.keys {
+		node, ok := g.keyToNode[key]
+		if !ok {
+			return fmt.Errorf("key not present: %v", key)
+		}
+		inboundEdges := make([]Key, len(node.inboundEdges))
+		copy(inboundEdges, node.inboundEdges)
+		outboundEdges := make([]Key, len(node.outboundEdges))
+		copy(outboundEdges, node.outboundEdges)
+		if err := f(key, inboundEdges, outboundEdges); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// WalkEdges visits each edge in the Graph starting at the source keys.
+//
+// f is called for each directed edge. The first argument is the source
+// node, the second is the destination node.
 //
 // Returns a *CycleError if there is a cycle in the graph.
-func (g *Graph[Key]) Walk(f func(Key, Key) error) error {
+func (g *Graph[Key]) WalkEdges(f func(Key, Key) error) error {
```

Comment:
This could be in the func signature instead of the doc

```suggestion
// f is called for each directed edge.
//
// Returns a *CycleError if there is a cycle in the graph.
func (g *Graph[Key]) WalkEdges(f func(sourceNode Key, destinationNode Key) error) error {
```